### PR TITLE
Button component and Design System

### LINKED ISSUE #3526: Final button component
## Description

This is an attempt to have a final single button component. It is also a pledge that we will not use any other button in new UI and in the designs.

We need a button that has:
- moves by 1px down on `:active` state (as most of them do already)
- three different types:
  1. bare
    - no background (rounder corners)
    - no border
    - becomes light gray background on hover
    - colored text
  2. frame
    - no background (rounder corners)
    - border 2px
    - becomes full color with white text on hover
    - colored text and border
  3. full
    - color background
    - no border
    - becomes a bit darker on hover
    - colored background
- three different sizes:
   1. `s` small - 28px height
   2. `m` medium - 32px height
   3. `l` large - 36px height
- multiple different colors
- each button can have a mix of type, size and color

We need the sizes to be configurable, as we're not 100% set on the ones mentioned above just yet.

## Additional details

Currently we have a dozen of buttons:
- KoboButton
- KoboLightButton
- KoboTextButton
- multiple custom buttons for mutiple different places in UI

We should deprecate existing components (a comment will be fine) and use new one from now on and migrate old stuff.

This is needed for #3415 to not write a new button again and again.

### PR DESCRIPTION
**Pull request by magicznyleszek**
## Description

Introduce a new Button component that covers all possible types, sizes and colors of buttons we will be using in UI.

## Additional details

Also introduces a `/design-system` (it's not `/#/design-system`!) route for displaying a "design system" - a demo of our components. Currently has `Icon`, the new `Button`, `KoboDropdown`, and a few more. The plan is to have all the common UI elements there in one place.

## Related issues

Includes code from #3579
Fixes #3526
Part of #3548

### CODE REVIEW

File: jsapp/js/designSystem/buttonDemo.tsx [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+      // Only one of icons is allowed.
+      demoStartIcon: newEndIcon ? null : this.state.demoStartIcon
+    })
+  }
+
+  onLabelChange(newLabel: string) {
+    this.setState({
+      demoLabel: newLabel,
+      // If there is no label, icon is required
+      demoStartIcon: newLabel === '' ? defaultIcon : this.state.demoStartIcon
```
Comment (duvld) [inline jsapp/js/designSystem/buttonDemo.tsx:97]: Not a big deal at all, but if we clear the text box then both icons in start and end show up:
Screenshot from 2022-01-03 16-05-26
looks like we would just not do this if `demoEndIcon` already exists

File: jsapp/js/designSystem/textboxDemo.tsx [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+import Radio from 'js/components/common/radio'
+
+const textBoxTypes: AvailableType[] = ['text-multiline', 'text', 'email', 'password', 'url', 'number']
+
+const errorsValues: {[val: string]: any} = {
+  no: null,
+  messageless: true,
+  one: 'You need to type something else!',
+  multi: [
+    "First error is about one thing!",
```
Comment (duvld) [inline jsapp/js/designSystem/textboxDemo.tsx:14]: Not a big deal at all, but if we clear the text box then both icons in start and end show up:
Screenshot from 2022-01-03 16-05-26

maybe a niche request but could we make these editable? may be useful to see how a responseJSON would look

File: jsapp/js/components/common/textBox.tsx [hunk 3]
```diff
@@ -3,19 +3,12 @@ import bem from 'js/bem';
...
-  text = 'text',
-  email = 'email',
-  password = 'password',
-  url = 'url',
-  number = 'number',
-}
+export type AvailableType = 'text-multiline' | 'text' | 'email' | 'password' | 'url' | 'number';
 
-const DefaultType = AvailableType.text;
+const DefaultType = 'text';
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/common/textBox.tsx:8]: I can't use it like this (i.e. as a value), as `AvailableType` is only type. But I fixed it by making it not a string: `const DefaultType: AvailableType = 'text'`

[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/common/textBox.tsx:8]: Also removed semicolons :)

File: jsapp/js/components/common/button.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    ) {
+      throw new Error(t('Button is missing a required properties: icon or label!'))
+    }
+
+    let classNames: string[] = []
+
+    // Additional class names.
+    if (
+      Array.isArray(this.props.classNames) &&
+      typeof this.props.classNames[0] === 'string'
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/common/button.tsx:96]: Yes :D

File: jsapp/js/components/common/button.scss [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  line-height: $height - 2 * $button-border-width; // Subtracting the border.
+  font-size: $font;
+
+  // Note: paddings and margins are based on the `$height` (the "size") variable
+  // to scale with button sizes.
+
+  // We don't want to add left & right paddings if there is only icon inside, as
+  // it means we are aiming at a square button.
+  &.k-button--has-label {
+    padding-left: $height * 0.4;
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/common/button.scss:86]: Trial and error :) I think I just made these based on the size from the designs for the medium size.

File: jsapp/js/designSystem/buttonDemo.tsx (hunk 1 above)
Comment (magicznyleszek) [reply jsapp/js/designSystem/buttonDemo.tsx:97]: Oh I see, in the `onLabelChange` function in `buttonDemo.tsx` I was setting a default icon when label was removed, but I was only setting one of the icons, with other remaining what was already there. Fixed that :) Plus for the `button.tsx` I am now displaying the end icon only if start icon wasn't defined.

File: jsapp/js/designSystem/textboxDemo.tsx (hunk 2 above)
Comment (magicznyleszek) [reply jsapp/js/designSystem/textboxDemo.tsx:14]: That would take some more time. But I did update the comment next to `errors` property in `textBox.tsx` to make it clear what you need to pass to component.

**Pull request merged**